torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -1911,6 +1911,58 @@ std::vector<For*> LoopNest::reorder(
   return result;
 }
 
+std::vector<For*> LoopNest::tile(For* x, For* y, int x_factor, int y_factor) {
+  auto parent = dynamic_cast<Block*>(x->get_parent());
+  if (parent == nullptr) {
+    throw malformed_input("parent of the loops must be a Block");
+  }
+  if (!areLoopsPerfectlyNested({x, y})) {
+    throw malformed_input("two loops must be perfectly nested");
+  }
+
+  // Split x, y axes by x_factor and y_factor
+  For *yi, *ytail;
+  splitWithTail(y, y_factor, &yi, &ytail);
+  For *xi, *xtail;
+  splitWithTail(x, x_factor, &xi, &xtail);
+
+  // Distribute xi over yo and ytail so we can manipulate the loop order of {xo,
+  // xi, yo, yi}
+  auto loops = distributeLoop(xi);
+
+  // For {xo, xi, yo, yi}, reorder the axes to be xo, yo, xi, yi
+  xi = loops.front();
+  For* yo = dynamic_cast<For*>(xi->body()->stmts().front());
+  CHECK(yo);
+  reorderAxis(xi, yo);
+
+  // Re-Identify loop handles of yo, xi, yi, ytail
+  For* xo = x;
+  yo = dynamic_cast<For*>(xo->body()->stmts().front());
+  CHECK(yo);
+  xi = dynamic_cast<For*>(yo->body()->stmts().front());
+  CHECK(xi);
+  yi = dynamic_cast<For*>(xi->body()->stmts().front());
+  CHECK(yi);
+
+  if (ytail) {
+    ytail = dynamic_cast<For*>(xo->body()->stmts().back());
+    CHECK(ytail);
+    ytail = dynamic_cast<For*>(ytail->body()->stmts().front());
+    CHECK(ytail);
+  }
+
+  std::vector<For*> result;
+  result.push_back(xo);
+  result.push_back(yo);
+  result.push_back(xi);
+  result.push_back(yi);
+  result.push_back(xtail);
+  result.push_back(ytail);
```

Comment:
Why do we have to return all the loops here? They could be at different level in nesting. It will be hard to make sense of it at the caller.

We could either return just the outer most loop, or since the given loops are valid after transformation, we can just return a bool to say that it is successful.

Reply:
As you noted, the tail loops are returned in awkward way. We can get around this by just returning a bool here to denote success. Since you are keeping this transform in-place, the higher-level API can extract the loops you need and return to the caller.

Reply:
Hmm, not for `xtail`. The user cannot extract the `xtail` loop because `xtail` is not nested inside `xo` (i.e., `x`, the only valid loop handle the user holds after `tile`). The user has to retrieve all loops for a buffer/tensor to find `xtail`.

Reply:
Or perhaps we can fuse the resulted loops in this way,
```
xo 
  yo
    xi
      yi
  ytail
    xi
xtail
```
returning `{xo, yo, xi, yi, ytail, xtail}` will make more sense: `ytail, xtail` are the root of the only two branches from the main loop and the user has all loop handles he/she is interested in.

Reply:
Good point regarding `xtail`. I agree that we have to find a way to return some of the transformed loops so that the user can retrieve the rest.

But returning `{xo, yo, xi, yi, ytail, xtail}` looks confusing to me. Why are we returning in that order? Why are we returning only one `xi`? Why not return the other `xi` (the one after distribute) as well? If we return a vector like this, is it okay to return a `nullptr` for `xtail` if no tail was generated? 

If you prefer to return all the transformed loops, here is one suggestion I have: return 3 different vectors each representing a loop nest in the transformed loop.
For example, for the transformed loop:
```
xo
  yo
    xi
      yi
  xi
    ytail
xtail
```
we return 3 vectors:
  * main loop nest: `xo, yo, xi, yi`
  * ytail loop nest: `xo, xi, ytail`
  * xtail loop nest: `xtail`
These can be 3 parameters to the API. The ytail and xtail loop nest parameters will be set to `nullptr` if the corresponding tails were not generated. What do you think of this? 

@ZolotukhinM  do you have any other idea for this?

Reply:
If we want to return all loops, I would prefer to return it as a vector of vectors, similar to how we return in `getAllLoopNestsFor`. That would convey the structure the transformation has made.

An alternative approach could be to put all these loops into a Block and return the block. This block would have a well-defined structure and the user would be able to easily locate necessary loops within it:
```cpp
Block* tile_stmt = tile(loop, 16, 16);
xo = tile_stmt.stmts()[0];
if (tile_stmt.stmts().size() > 1) 
  xtail = tile_stmt.stmts()[1];
else
  xtail = nullptr;
yo = ...;
```
This code of course could be wrapped in a nicer looking API if needed.

The drawback of this approach is that it is not used anywhere else and can be surprising.

Reply:
> If we want to return all loops, I would prefer to return it as a vector of vectors, similar to how we return in getAllLoopNestsFor. That would convey the structure the transformation has made.

The reason I wasn't suggesting this is because we want to have all our transforms return a `bool`. If it is going to be an out parameter, why not have 3 out-parameters :)

Reply:
> An alternative approach could be to put all these loops into a Block and return the block.

The loops will already be part of a parent block. So, we can't add them to a new block. If we return the parent block, it is not clear how to identify `xtail`.

Reply:
> The loops will already be part of a parent block. So, we can't add them to a new block. If we return the parent block, it is not clear how to identify xtail.

Yes, what I meant is to construct a nested Block just for the statements from the tile transform - that block would be a statement in the initial loop's parent block. IRSimplify would get rid of that block once applied, but before that we can use it to identify loops.

Reply:
> The reason I wasn't suggesting this is because we want to have all our transforms return a bool.

Why? Are we going to have `splitWithTail` not returning `xi, xtail` but a bool? @navahgar

Reply:
`splitWithTail` returns `xi` and `xtail` as out-parameters. That will stay the same. But all transformations should return a bool indicating success. See: https://github.com/pytorch/pytorch/issues/56357

Reply:
> An alternative approach could be to put all these loops into a Block and return the block. This block would have a well-defined structure and the user would be able to easily locate necessary loops within it

I am thinking of a similar approach. While returning the main loop `xo, yo, xi, yi` and the roots of tailing loops `ytail, xtail` in `tile`, we also provide another API, say `getLoopsIn()` which returns loop vectors in a root loop stmt. The users are able to retrieve all loop vectors in a tailing loop if they'd like to perform more transformations on them. Otherwise, they already have all loop handles in the main loop.

Reply:
>But returning {xo, yo, xi, yi, ytail, xtail} looks confusing to me. Why are we returning in that order? 

It matches where they appear in the transformed loop:
```
xo
  yo
    xi
      yi
  ytail
xtail
```  
`xtail, ytail` will be `nullptr` if there is no such corresponding tailing loop.

Reply:
Why are we not returning the other `xi` under `ytail`?

Reply:
We are returning `ytail` and the user can use another API `getLoopIn()` to retrieve `xi`. Same for `xtail`, below is the structure of `xtail`:
```
xtail
  yo
    yi
  ytail
```
The user is able to retrieve all loop vectors in `xtail` by passing `xtail` handle to `getLoopsIn()`.

This way we speedup the path to main loop transformation as the user holds all loop handles in the main loop; but also support transformations on tail loops as we return their roots and provide an API to retrieve all loops in them.

Reply:
Don't we generate the following structure?
```
xo
  yo
    xi
      yi 
  xi 
    ytail
xtail
```
i.e., `ytail` inside `xi` and not the other way around.

Given this and the logic that we should be returning the start of each loop, shouldn't we return `xi` instead of `ytail`?

Reply:
> splitWithTail returns xi and xtail as out-parameters. That will stay the same. 

`tile` transformation seems similar to `split` transformation to me. Both generates new loop axes and should return them to the user. Also, we would never return `Flase` if we return a bool in `tile`: say `x, y` are not perfectly nested, we should throw an exception because that's a pre-condition to tile; and tile two perfectly nested loops would never fail. If the specified factors are both larger than their loop bounds (a corner case I should add in the tests :smile: ), we would not touch the original stmt. it, however, an be considered as the simplified stmt for a tiled version.

Reply:
> Given this and the logic that we should be returning the start of each loop, shouldn't we return xi instead of ytail?

Ah, that's what I propose to change too : we reorder `xi` and `ytail` so the new order `ytail, xi` kinda matches the main loop `yo xi`; and this way `ytail` becomes the root of the tail loop.

Reply:
In summary, my solution is to return `xo, yo, xi, yi, ytail, xtail` as in the following structure:
```
xo
  yo
    xi
      yi
  ytail
    ytail-xi
xtail
  xtail-yo
    xtail-yi
  xtail_ytail
```
, and provide an API `getLoopsIn()` to retrieve all loop vectors in a root loop (for further transformations on tail loops `xtail, ytail` if any).

Reply:
> say x, y are not perfectly nested, we should throw an exception because that's a pre-condition to tile;

We want to avoid throwing exceptions in such cases since that is still a valid IR, just an invalid input to this transformation. It is difficult when exceptions are thrown because not all callers catch exceptions, so, the process gets killed completely. Hence this issue: #56357